File: src/execution_context/FExecutionContext.ts
import { FCancellationToken } from "../cancellation/FCancellationToken";

import { FExceptionInvalidOperation } from "../exception/FExceptionInvalidOperation";

export abstract class FExecutionContext {
	public abstract get prevContext(): FExecutionContext | null;

	/**
	 * Provide empty execution context. Usually used as root of execution context chain.
	 */
	public static get Empty(): FExecutionContext {
		return _EmptyExecutionContext.instance;
	}

	/**
	 * Provide default execution context.
	 *
	 * The execution context contains:
	 * * `FCancellationExecutionContext` with `FCancellationToken.Dummy`
	 */
	public static get Default(): FExecutionContext {
		return _DefaultExecutionContext.instance;
	}

	/**
	 * Obtain a closest instance of typed `FExecutionContext` that encloses
	 * the given context.
	 *
	 * Returns `null` if requested type not found
	 */
	protected static findExecutionContext<T extends FExecutionContext>(
		context: FExecutionContext, clz: Function & { prototype: T; }
	): T | null {
		let chainItem: FExecutionContext | null = context;
		while (chainItem !== null) {
			if (chainItem instanceof clz) {
				return chainItem as T;
			}
			chainItem = chainItem.prevContext;
		}
		return null;
	}

	protected static findAllExecutionContexts<T extends FExecutionContext>(
		context: FExecutionContext, clz: Function & { prototype: T; }
	): ReadonlyArray<T> {
		const result: Array<T> = [];

		let chainItem: FExecutionContext | null = context;
		while (chainItem != null) {
			if (chainItem instanceof clz) {
				result.push(chainItem as T);
			}
			chainItem = chainItem.prevContext;
		}

		return Object.freeze(result);
	}

	/**
	 * Obtain a closest instance of typed `FExecutionContext` that encloses
	 * the given context.
	 *
	 * Raise `FExceptionInvalidOperation` if requested type not found
	 */
	protected static getExecutionContext<T extends FExecutionContext>(context: FExecutionContext, clz: Function & { prototype: T; }): T {
		const chainItem: T | null = FExecutionContext.findExecutionContext(context, clz);
		if (chainItem !== null) {
			return chainItem;
		}
		throw new FExceptionInvalidOperation(`Execution context '${clz.name}' is not presented on the chain.`);
	}
}
export class FExecutionElement<TExecutionContext extends FExecutionContext> {
	private readonly _owner: TExecutionContext;

	public constructor(owner: TExecutionContext) {
		this._owner = owner;
	}

	public get owner(): TExecutionContext { return this._owner; }
}

// Import here due to cyclic dependencies
import { FCancellationExecutionContext } from "../cancellation/FCancellationExecutionContext";

class _DefaultExecutionContext extends FExecutionContext {
	private static _instance: _DefaultExecutionContext | null = null;
	public static get instance(): _DefaultExecutionContext {
		if (_DefaultExecutionContext._instance === null) {
			_DefaultExecutionContext._instance = new _DefaultExecutionContext();
		}
		return _DefaultExecutionContext._instance;
	}

	private readonly _prevContext: FExecutionContext;

	public constructor() {
		super();

		this._prevContext = new FCancellationExecutionContext(
			_EmptyExecutionContext.instance,
			FCancellationToken.Dummy,
		);
	}

	public get prevContext(): FExecutionContext | null { return this._prevContext; }
}

class _EmptyExecutionContext extends FExecutionContext {
	private static _instance: _EmptyExecutionContext | null = null;
	public static get instance(): _EmptyExecutionContext {
		if (_EmptyExecutionContext._instance === null) {
			_EmptyExecutionContext._instance = new _EmptyExecutionContext();
		}
		return _EmptyExecutionContext._instance;
	}

	public get prevContext(): FExecutionContext | null { return null; }
}

File: lib/execution_context/f_execution_context.dart
import "package:meta/meta.dart";

import "../cancellation/f_cancellation_execution_context.dart"
    show FCancellationExecutionContext;
import "../cancellation/f_cancellation_token.dart" show FCancellationToken;
import "../exception/f_exception_invalid_operation.dart"
    show FExceptionInvalidOperation;

class FExecutionContext {
  final FExecutionContext? prevContext;

  const FExecutionContext(this.prevContext);

  ///
  /// Provide empty execution context. Usually used as root of execution context chain.
  ///
  static FExecutionContext get emptyExecutionContext {
    return _EmptyExecutionContext.instance;
  }

  ///
  /// Provide default execution context.
  ///
  /// The execution context contains:
  /// * `FCancellationExecutionContext` with `FCancellationToken.Dummy`
  ///
  static FExecutionContext get defaultExecutionContext {
    return _DefaultExecutionContext.instance;
  }

  ///
  /// Obtain a closest instance of typed `FExecutionContext` that encloses
  /// the given context.
  ///
  /// Returns `null` if requested type not found
  ///
  @protected
  static T? findExecutionContext<T extends FExecutionContext>(
    final FExecutionContext context,
  ) {
    FExecutionContext? chainItem = context;
    while (chainItem != null) {
      if (chainItem is T) {
        return chainItem;
      }
      chainItem = chainItem.prevContext;
    }
    return null;
  }

  ///
  /// Obtain a closest instance of typed `FExecutionContext` that encloses
  /// the given context.
  ///
  /// Raise `FExceptionInvalidOperation` if requested type not found
  ///
  @protected
  static T getExecutionContext<T extends FExecutionContext>(
    final FExecutionContext context,
  ) {
    final T? chainItem = FExecutionContext.findExecutionContext<T>(context);

    if (chainItem != null) {
      return chainItem;
    }

    throw FExceptionInvalidOperation(
        "Execution context '$T' is not presented on the chain.");
  }

  static List<T> listExecutionContexts<T extends FExecutionContext>(
    final FExecutionContext context,
  ) {
    final List<T> result = [];

    FExecutionContext? chainItem = context;
    while (chainItem != null) {
      if (chainItem is T) {
        result.add(chainItem);
      }
      chainItem = chainItem.prevContext;
    }

    return List.unmodifiable(result);
  }
}

class FExecutionElement<TExecutionContext extends FExecutionContext> {
  final TExecutionContext _owner;

  FExecutionElement(TExecutionContext owner) : _owner = owner;

  TExecutionContext get owner {
    return this._owner;
  }
}

class _DefaultExecutionContext implements FExecutionContext {
  static _DefaultExecutionContext? _instance;
  static _DefaultExecutionContext get instance {
    // ignore: prefer_conditional_assignment
    if (_DefaultExecutionContext._instance == null) {
      _DefaultExecutionContext._instance = _DefaultExecutionContext();
    }
    return _DefaultExecutionContext._instance!;
  }

  final FExecutionContext _prevContext;

  _DefaultExecutionContext()
      : _prevContext = FCancellationExecutionContext(
          _EmptyExecutionContext.instance,
          FCancellationToken.dummy,
        );

  @override
  FExecutionContext? get prevContext {
    return this._prevContext;
  }
}

class _EmptyExecutionContext implements FExecutionContext {
  static _EmptyExecutionContext? _instance;
  static _EmptyExecutionContext get instance {
    // ignore: prefer_conditional_assignment
    if (_EmptyExecutionContext._instance == null) {
      _EmptyExecutionContext._instance = _EmptyExecutionContext();
    }
    return _EmptyExecutionContext._instance!;
  }

  @override
  FExecutionContext? get prevContext => null;
}
